KafkaStore SSL support.

alexlod opened the pull request:
This diff adds support for an SSL KafkaStore. The implementation is pretty straightforward, mostly involving forwarding config parameters to the Kafka clients. The tests were a little tricky, though.

There were a few implementation choices I made which I'm not sure are ideal:

 * Some Kafka client configuration parameters have a default value of `null`. `ConfigDef` doesn't support optional parameters with a default value of `null`. My workaround is to make the default value in the schema registry `""` and convert that to `null` when I initialize the Kafka clients.
 * I introduced two new test classes, one testing SSL with authentication, and the other testing SSL without authentication. I copy-pasted some code between these two, which feels bad. However, I couldn't figure out a better alternative.

Thanks ahead of time for taking a look!

ewencp commented on bin/schema-registry-run-class:
```diff
@@ -63,7 +63,7 @@ fi
 # Memory options
 if [ -z "$SCHEMA_REGISTRY_HEAP_OPTS" ]; then
-  SCHEMA_REGISTRY_HEAP_OPTS="-Xmx256M"
+  SCHEMA_REGISTRY_HEAP_OPTS="-Xmx512M"
```
This seems quite large for what the schema registry is doing. Did you actually run into limits with the existing setting?

alexlod replied:
@ewencp, yeah I was getting OOMEs. I didn't test whether a smaller value would work or not. Want me to test a smaller value, like `384`?

ewencp commented on core/pom.xml:
```diff
@@ -76,6 +76,18 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>org.apache.kafka</groupId>
```
Doesn't this already get pulled in via the `kafka_${kafka.scala.version}` dependency?

alexlod replied:
No, tests won't compile without this dependency. It's needed for the use of `TestSslUtils`.

ewencp replied:
Ack, was misreading this as the normal, not test, jar.

ewencp commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -76,6 +79,28 @@
    * <code>avro.compatibility.level</code>
    */
   public static final String COMPATIBILITY_CONFIG = "avro.compatibility.level";
+  public static final String KAFKASTORE_SECURITY_PROTOCOL_CONFIG =
```
Can we leverage the client security config support in Kafka to simplify a bunch of this (see `ConfigDef.withClientSslSupport`/`ConfigDef.withClientSaslSupport`)? Not sure it's possible given the additional prefix, but it seems easier to maintain and less error prone if we can just get the configs directly from Kafka instead of copy/pasting here.

alexlod replied:
I didn't realize these existed -- great! I agree that this may be a better route. I'll experiment with prepending `kafkastore.` to each.

alexlod replied:
This change would require we use Kafka's `ConfigDef`, which we decided we wouldn't do now. So although it's a great idea, for now we can't do it.

hachikuji replied:
Maybe add a TODO if it's something we think we'll do eventually?

Ishiihara replied:
@alexlod Any particular reason we are not using Kafka's `ConfigDef`?

Ishiihara replied:
The reason it because we don't have `defineOverride`, correct?

alexlod replied:
I'll experiment with @Ishiihara's suggestion below, to change `ConfigDef` to support null defaults. Regardless, I'll add a TODO per @hachikuji's suggestion.

ewencp commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -151,6 +207,15 @@ public void init() throws StoreInitializationException {
+  // because some Kafka client configs expect a null default value, yet ConfigDef doesn't
+  // support null default values.
+  public static void putIfNotEmptyString(String parameter, String value, Properties props) {
+    if (! value.equals("")) {
```
you might want to `trim()` the string, and `isEmpty()` is probably clearer.

alexlod replied:
Good suggestions!

ewencp commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -73,7 +76,18 @@ public KafkaStoreReaderThread(String bootstrapBrokers,
                                 Store<K, V> localStore,
-                                K noopKey) {
+                                K noopKey,
+                                String kafkastoreSecurityProtocol,
```
Hmm, not sure about this approach to passing in the configs -- seems like an excessive number of arguments. Any reason we can't just pass around the original configs map and pull out exactly the set of configs we need? The current approach seems like it could get especially messy if we add support for anything besides SSL.

alexlod replied:
Agreed. And it's tedious to code! I'll look into passing the config map.

ewencp commented on core/src/test/java/io/confluent/kafka/schemaregistry/SSLClusterTestHarness.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2014 Confluent Inc.
```
Heh, the date might need updating here.

alexlod replied:
I wasn't sure if we keep the founding year of Confluent, or this year. I'll change to this year.

ewencp commented:
@alexlod For the first `ConfigDef` issue, is this a problem with the Confluent version of those classes that's been fixed in Kafka trunk? I'm wondering if this can be addressed just by switching to the Apache classes? We're making `ConfigDef` public API in the next release since Connectors will use it to specify configuration info.

For the two new classes, can enough be saved by sharing a base class to make adding yet one more class into the mix?

alexlod commented:
@ewencp, great suggestions, thanks! I just pushed a new commit with trivial changes. I'll work on the less trivial suggestions now and report back when I have a new commit. Thanks again.

alexlod commented:
@ewencp and I spoke in person, and I'm documenting the result here. `ConfigDef` in Kafka doesn't have `defineOverride`. Given this, there are two approaches to use Kafka's `ConfigDef` in the `rest-utils` repo: modify Kafka's `ConfigDef` to have `defineOverride` which would require a KIP; or remove `RestConfig.baseConfigDef` and instead define these defaults in each REST application manually. The latter feels annoying, so we decided the best solution is the former. But considering we'd need a KIP and an upstream Kafka change, in this patch, we'll keep the workarounds and change them later when `ConfigDef` has `defineOverride`.

alexlod commented:
@ewencp: new commit addressing all your suggestions (minus the `ConfigDef` ones). I also added one more test requiring client auth but not providing client certs should cause a TimeoutException.

Thanks, and let me know what other suggestions you have.

hachikuji commented on core/pom.xml:
```diff
@@ -76,6 +76,18 @@
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcpkix-jdk15on</artifactId>
+            <version>1.54</version>
```
Nitpick: should we add this as a property in the parent POM?

alexlod replied:
Good catch.

hachikuji commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreReaderThread.java:
```diff
@@ -95,6 +99,32 @@ public KafkaStoreReaderThread(String bootstrapBrokers,
+
+    consumerProps.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG,
+            config.getString(SchemaRegistryConfig.KAFKASTORE_SECURITY_PROTOCOL_CONFIG));
+    if (config.getString(SchemaRegistryConfig.KAFKASTORE_SECURITY_PROTOCOL_CONFIG).equals(
```
Is there any difference between this block and the same one for the producer?

Ishiihara replied:
There maybe a null pointer exception in `config.getString` call.

alexlod replied:
@hachikuji, no. I'll make a helper method. Good catch -- newb mistake!

@Ishiihara, each config option has a default value, which I believe means the String key will never be missing from the config. Unless I'm missing something?

hachikuji commented on core/src/test/java/io/confluent/kafka/schemaregistry/ClusterTestHarness.java:
```diff
@@ -144,14 +133,33 @@ public void setUp() throws Exception {
+            TestUtils.RandomPort(), Option.<String>empty());
+    props.setProperty("auto.create.topics.enable", "true");
+    props.setProperty("num.partitions", "1");
+    // We *must* override this to use the port we allocated (Kafka currently allocates one port
```
Not sure I follow this comment. What does it mean for Kafka to allocate a port for Zookeeper?

hachikuji replied:
Sorry, just notice that this was pasted from the original code. Still might be worth clarifying this comment.

alexlod replied:
No worries. Will do.

hachikuji commented on core/src/test/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreSSLAuthTest.java:
```diff
@@ -0,0 +1,94 @@
+    try {
+      kafkaStore.put(key, value);
+    } catch (StoreException e) {
+      fail("Kafka store put(Kafka, Rocks) operation failed");
```
Instead of failing, maybe you can just propagate the exception? Otherwise, we won't be able to see the stacktrace in the output.

alexlod replied:
Good suggestion. I'll change this in `KafkaStoreTest` too.

Ishiihara commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -151,6 +183,15 @@ public void init() throws StoreInitializationException {
+  // because some Kafka client configs expect a null default value, yet ConfigDef doesn't
+  // support null default values.
+  public static void putIfNotEmptyString(String parameter, String value, Properties props) {
+    if (! value.trim().isEmpty()) {
```
nit: space between `!` and `value`

alexlod replied:
Fixed!

hachikuji commented on core/src/test/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreSSLAuthTest.java:
```diff
@@ -0,0 +1,94 @@
+  }
+
+  @Test
+  public void testIncorrectInitialization() {
```
Can you make the test name more specific? Are there different cases to check for incorrect initialization?

alexlod replied:
Good suggestion. I made the change in `KafkaStoreTest` too.

Ishiihara commented on core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java:
```diff
@@ -151,6 +183,15 @@ public void init() throws StoreInitializationException {
+  // helper method to only add a property if its not the empty string. This is required
+  // because some Kafka client configs expect a null default value, yet ConfigDef doesn't
+  // support null default values.
+  public static void putIfNotEmptyString(String parameter, String value, Properties props) {
```
This PR enables null default value https://github.com/apache/kafka/pull/936.

alexlod replied:
@Ishiihara, I've been experimenting with this, and unless I'm missing something, the changes in that PR don't solve my issue. My issue is that I want a null default value, not an empty string default value.

I modified `ConfigDef` to support my use case. Take a look: https://github.com/alexlod/common/commit/eea951b01653ab4abfc89d262e870392ed7b8ed2

I'm not comfortable committing the changes as they stand, because I noticed some odd behavior in my `testNullDefaultValueWithValidator()` test. Namely, if a "null" validator is used, the test passes. However, if a validator is not used, the test fails, because the actual value is `""`. I expect I could fix this, but I wonder if it's worth it considering the `ConfigDef` changes are non-trivial.

Anyway, take a look, and let me know what you think.

Ishiihara commented on docs/security.rst:
```diff
@@ -0,0 +1,12 @@
+
+Security Overview
+-----------------
+The Schema Registry currently only supports communication with a secure Kafka cluster over SSL. At this time, REST API and ZooKeeper security is not yet supported. Kafka SASL authentication is not supported yet either.
```
REST API is a bit confusing here. May be we should say the only support request over http not https

alexlod replied:
Good idea. Changed!

alexlod commented:
New commit addressing all comments (exception the `ConfigDef` ones discussed above). Also fixing a merge conflict.

ijuma commented on core/pom.xml:
```diff
@@ -82,6 +82,18 @@
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcpkix-jdk15on</artifactId>
+            <version>${bouncycastle.bcpkix.version}</version>
```
We are using bouncycastle outside of tests?

alexlod replied:
I don't believe so. I'll make the `scope` `test` and test on my local machine. Good catch!

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -152,7 +205,40 @@
+                        METRICS_JMX_PREFIX_DOC)
+        .define(KAFKASTORE_SECURITY_PROTOCOL_CONFIG, ConfigDef.Type.STRING,
+                KAFKASTORE_SECURITY_PROTOCOL_PLAINTEXT, ConfigDef.Importance.HIGH,
+                KAFKASTORE_SECURITY_PROTOCOL_DOC)
```
The importance of this is `MEDIUM` for the producer and consumer config, is it intentional that we set it to `HIGH` here? It probably makes sense, thinking about it, but just wanted to double-check.

ijuma replied:
The importance for a few other configs has also been changed. It would be interesting to have an explanation of the reasoning (I don't doubt that it's sound :)).

ijuma replied:
By the way, if we agree on the changes in priority, then you can submit a PR to change it in Kafka too.

alexlod replied:
I made up my own priorities based on what felt good. I don't feel strongly about keeping them as is vs being consistent with Kafka. I'll make them consistent with Kafka unless I hear otherwise.

ijuma replied:
I'm fine either way, I think thinking through priorities would probably be a good thing, but given our tight deadline, it probably makes sense to just do what Kafka does.

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -76,6 +79,28 @@
+      "kafkastore.ssl.keystore.password";
+  public static final String KAFKASTORE_SSL_KEY_PASSWORD_CONFIG =
+      "kafkastore.ssl.key.password";
+  public static final String KAFKASTORE_SSL_ENABLED_PROTOCOL_CONFIG =
```
Should be `PROTOCOLS` with an `S` at the end. The doc constant also needs to be updated.

alexlod replied:
Fixed. Good catch!

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -76,6 +79,28 @@
+      "kafkastore.ssl.protocol";
+  public static final String KAFKASTORE_SSL_PROVIDER_CONFIG =
+      "kafkastore.ssl.provider";
+  public static final String KAFKASTORE_SSL_TRUSTSTORE_TYPE_CONFIG =
```
Nitpick: it would be a bit nicer if all related configs were together (eg all truststore configs were grouped, all keystore configs were grouped, etc.)

alexlod replied:
I kept them in the same order as they are in the Kafka docs. But I'll group them. Good suggestion.

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -112,9 +137,37 @@
+  protected static final String KAFKASTORE_SSL_KEY_PASSWORD_DOC =
+      "The password of the key contained in the keystore configured with "
+      + "`kafka.ssl.keystore.location`.";
+  public static final String KAFAKSTORE_SSL_ENABLED_PROTOCOL_DOC =
```
As stated earlier, should be `PROTOCOLS` with an `S` at the end.

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -112,9 +137,37 @@
       + "full (new schema is backward and forward compatible with latest registered schema)";
   protected static final String MASTER_ELIGIBILITY_DOC = 
       "If true, this node can participate in master election. In a multi-colo setup, turn this off" 
```
Looks like we need a space after `off` (I know this is existing code).

alexlod replied:
Fixed!

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -112,9 +137,37 @@
+      "The password to access the keystore configured with `kafka.ssl.keystore.location`.";
+  protected static final String KAFKASTORE_SSL_KEY_PASSWORD_DOC =
+      "The password of the key contained in the keystore configured with "
+      + "`kafka.ssl.keystore.location`.";
```
Is it really useful to cross-reference the configs in this way? It seems to me that `The password of the private key contained in the keystore` (i.e. how it's done in the Kafka code) is fine (the location is a property of the keystore in the same way as the password). I'd be interested in what others think though.

ijuma replied:
Another reason for not doing it this way is that one doesn't have to configure the location at all if the certificates are included in the JDK (see https://github.com/apache/kafka/pull/312 for more background).

alexlod replied:
Will remove references, both in the code and in the docs.

ijuma commented on docs/config.rst:
```diff
@@ -51,6 +51,13 @@ Configuration Options
   * Default: "localhost"
   * Importance: high
 
+``kafkastore.security.protocol``
```
Can you please also update https://github.com/confluentinc/docs/blob/master/kafka/platform-security.rst once we merge this?

alexlod replied:
Will do. Thanks for the heads up -- I didn't know that doc existed!

alexlod replied:
Here's a pull request for this change: https://github.com/confluentinc/docs/pull/103

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -152,7 +205,40 @@
+                "", ConfigDef.Importance.MEDIUM,
+                KAFKASTORE_SSL_KEY_PASSWORD_DOC)
+        .define(KAFKASTORE_SSL_ENABLED_PROTOCOL_CONFIG, ConfigDef.Type.STRING,
+                "", ConfigDef.Importance.LOW,
```
What's the reasoning for leaving it undefined instead of Kafka's default ("TLSv1.2,TLSv1.1,TLSv1")?

alexlod replied:
I was trying `[TLSv1.2,TLSv1.1,TLSv1]` as listed in the Kafka docs but was getting an error (forget which error). I'll try without the square brackets.

ijuma commented on core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java:
```diff
@@ -152,7 +205,40 @@
+        .define(KAFKASTORE_SSL_PROVIDER_CONFIG, ConfigDef.Type.STRING,
+                "", ConfigDef.Importance.LOW,
+                KAFAKSTORE_SSL_PROVIDER_DOC)
+        .define(KAFKASTORE_SSL_TRUSTSTORE_TYPE_CONFIG, ConfigDef.Type.STRING,
```
A few configs are missing (cipher suites, key manager/trust manager algorithm, endpoint identification...), what is the reason for that?

alexlod replied:
I somehow missed them. I'll add them and make sure I didn't miss others. Good catch.

granders commented on core/src/test/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreTest.java:
```diff
@@ -83,18 +83,23 @@ public void testSimplePut() throws InterruptedException {
+        throw new RuntimeException("Kafka store get(Kafka) operation failed", e);
+      }
+      assertEquals("Retrieved value should match entered value", value, retrievedValue);
+    } catch (RuntimeException re) {
```
Looks like this outer try can just be try/finally without the catch?

alexlod replied:
You're right! Good catch.

granders commented on core/src/test/java/io/confluent/kafka/schemaregistry/storage/KafkaStoreTest.java:
```diff
@@ -137,30 +142,39 @@ public void testSimpleGetAfterFailure() throws InterruptedException {
+        throw new RuntimeException("Kafka store get(Kafka) operation failed", e);
+      }
+      assertEquals("Retrieved value should match entered value", value, retrievedValue);
+    } catch (RuntimeException re) {
```
Minor, and same as above - don't really need this catch?

alexlod replied:
Fixed! And fixed the other occurrences, too.

granders commented:
@alexlod Thanks for this update! I had only extremely minor comments.

A couple questions which probably don't impact this patch
- as I understand, this provides encryption between schema registry and kafka. Is the plan also to ultimately support https for the REST interface?
- It may be good to add at least one simple system/end to end test to muckrake - were you planning anything along these lines?

alexlod commented:
@ijuma and @granders, thanks for your feedback! All great suggestions! I'll work on them and send out another comment when I'm done.

@granders, to answer your question, yes, the ultimate goal is to do https and ZooKeeper SASL. I'm nearly finished with the https part and will hopefully have a patch later today. Then, I'll start on ZooKeeper SASL. As for muckrake tests, my thought was to make a system test that tested all Schema Registry security aspects, not just the Kafka SSL component. And I thought I'd write that test when I finish all the security features. Is that ok? Or would you rather I write the test now?

granders commented:
@alexlod Cool your plan sounds good

alexlod commented:
@ijuma and @granders, I pushed a rebased commit with all your suggestions but this pull request isn't updating for some reason. The commit is here:

https://github.com/alexlod/schema-registry/commit/e2ec9c6f800fff60d68fef42cd4f6708c6c204fd

Want me to close this pull request and open a new one?

ewencp commented:
@alexlod That looks like its on the branch `kafka-ssl`, this PR is tied to `kafakstore-ssl`. You'll need to push to the same branch to get this PR to update.

alexlod commented:
@ewencp, good catch, (facepalm). This PR is now updated.

ijuma commented:
I had a brief look and it looks fine to me. For the future, while review is still taking place it's a bit nicer if the changes are done via additional commits so that one can see the changes since the last review. It's fine to squash commits before merging the PR in order to make the history easier to follow.

alexlod commented:
Good feedback, @ijuma! Will do this in the future. Thanks.

ewencp commented:
![yeah](https://media.giphy.com/media/4dJ2ekKoBZq6Y/giphy.gif)
LGTM. Thanks for all the iterations on this @alexlod!

alexlod commented:
Totally, @ewencp! It was fun! And I'm very grateful for everyone's great suggestions.
